pkg/project/auth/cache.go

```diff
@@ -301,21 +246,27 @@ func purgeDeletedNamespaces(namespaceSet *util.StringSet, userSubjectRecordStore
 func (ac *AuthorizationCache) invalidateCache() bool {
 	invalidateCache := false
 
-	clusterPolicies := ac.clusterPolicyStore.List()
-	for i := range clusterPolicies {
-		policy := clusterPolicies[i].(*authorizationapi.ClusterPolicy)
-		if policy.ResourceVersion != ac.clusterPolicyResourceVersion {
+	clusterPolicyList, err := ac.policyClient.ReadOnlyClusterPolicies().List(labels.Everything(), fields.Everything())
+	if err != nil {
+		util.HandleError(err)
+		return invalidateCache
+	}
+	for _, clusterPolicy := range clusterPolicyList.Items {
+		if clusterPolicy.ResourceVersion != ac.clusterPolicyResourceVersion {
 			invalidateCache = true
-			ac.clusterPolicyResourceVersion = policy.ResourceVersion
+			ac.clusterPolicyResourceVersion = clusterPolicy.ResourceVersion
 		}
 	}
 
-	clusterPolicyBindings := ac.clusterPolicyBindingStore.List()
-	for i := range clusterPolicyBindings {
-		policyBinding := clusterPolicyBindings[i].(*authorizationapi.ClusterPolicyBinding)
-		if policyBinding.ResourceVersion != ac.clusterBindingResourceVersion {
+	clusterPolicyBindingList, err := ac.policyClient.ReadOnlyClusterPolicyBindings().List(labels.Everything(), fields.Everything())
+	if err != nil {
+		util.HandleError(err)
+		return invalidateCache
+	}
+	for _, clusterPolicyBinding := range clusterPolicyBindingList.Items {
+		if clusterPolicyBinding.ResourceVersion != ac.clusterBindingResourceVersion {
 			invalidateCache = true
-			ac.clusterBindingResourceVersion = policyBinding.ResourceVersion
+			ac.clusterBindingResourceVersion = clusterPolicyBinding.ResourceVersion
```

Comment:
then prior to exiting this function update ac.clusterBindingResourceVersion and ac.clusterPolicyResourceVersion sets with the latest observed set.